File: C++/Bucket_Sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void insertionSort(vector<double>& bukt) {
    for (size_t j = 1; j < bukt.size(); j++) {
        double val = bukt[j];
        int k = j - 1;
        while (k >= 0 && bukt[k] > val) {
            bukt[k + 1] = bukt[k];
            k--;
        }
        bukt[k + 1] = val;
    }
}

void bucketSort(vector<double>& arr) {
    int s = arr.size();
    vector<vector<double>> buckets(s);

    for (double num : arr) {
        int bi = int(s * num);
        if (bi >= s) bi = s - 1;
        buckets[bi].push_back(num);
    }

    for (auto& b : buckets) {
        insertionSort(b);
    }

    int idx = 0;
    for (auto& b : buckets) {
        for (double num : b) {
            arr[idx++] = num;
        }
    }
}

int main() {
    vector<double> inputArr = {0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67};
    cout << "Arreglo antes de ordenar:\n";
    for (double x : inputArr) cout << x << " ";
    cout << endl;

    bucketSort(inputArr);

    cout << "Arreglo después de ordenar:\n";
    for (double x : inputArr) cout << x << " ";
    cout << endl;
}

File: Java/Bucket_Sort.java
import java.util.List;
import java.util.ArrayList;

public class Bucket_Sort {

    // Método de ordenamiento por inserción para cada cubeta
    static void insertionSort(List<Double> bukt) {
        for (int j = 1; j < bukt.size(); j++) {
            double val = bukt.get(j);
            int k = j - 1;
            while (k >= 0 && bukt.get(k) > val) {
                bukt.set(k + 1, bukt.get(k));
                k--;
            }
            bukt.set(k + 1, val);
        }
    }

    // Método principal del algoritmo Bucket Sort
    static void bucketSort(double[] arr) {
        int s = arr.length;
        @SuppressWarnings("unchecked")
        List<Double>[] buckets = (List<Double>[]) new List[s]; // CORREGIDO
        for (int i = 0; i < s; i++)
            buckets[i] = new ArrayList<>();

        // Distribuir los elementos en cubetas
        for (double num : arr) {
            int bi = (int)(s * num);
            if (bi >= s) bi = s - 1;
            buckets[bi].add(num);
        }

        // Ordenar cada cubeta
        for (List<Double> b : buckets)
            insertionSort(b);

        // Combinar las cubetas
        int idx = 0;
        for (List<Double> b : buckets)
            for (double num : b)
                arr[idx++] = num;
    }

    public static void main(String[] args) {
        double[] inputArr = {0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67};

        System.out.println("Arreglo antes de ordenar:");
        for (double x : inputArr) System.out.print(x + " ");
        System.out.println();

        bucketSort(inputArr);

        System.out.println("Arreglo después de ordenar:");
        for (double x : inputArr) System.out.print(x + " ");
        System.out.println();
    }
}